### ISSUE TITLE
[Bug] DeepSeek V4 greedy MTP decode diverges from AR token output

### ISSUE THREAD

**Issue by vegetabledoww** [hw-native-sys/pypto-serving#161]
## Diagnosis

This appears to be in the DeepSeek V4 greedy MTP target-decode path at the
`pypto-serving` / `pypto-lib` boundary. The host publication path has one
separate committed-window accounting issue, but fixing that only postpones the
visible divergence: the S=2 target row 0 logits still differ enough from S=1 AR
to change the greedy token.

## Description

DeepSeek V4 Flash W8A8 produces a different greedy token sequence when K=1 MTP
is enabled, even though the prompt, committed token history, sampling settings,
and cache configuration are otherwise identical.

Expected behavior: greedy speculative decoding is lossless. With
`temperature=0`, enabling MTP may change execution and the number of accepted
drafts, but it must not change the target model's committed token sequence.

Actual behavior:

- Synchronous MTP runs are repeatable and return the same sequence, so this is
  not per-run sampling randomness.
- Async and sync MTP both reproduce the problem; async scheduling is not the
  initial numerical cause.
- The default 16-row MTP tile and a temporary 8-row MTP tile both reproduce it;
  decode-tile expansion is not the root cause.
- A local fix that publishes the canonical committed-window suffix instead of
  `sampled_ids[:accepted_count]` moves the first visible divergence from index
  21 to index 48, but does not restore target-logit equivalence.
- At the remaining divergence, the first 48 generated tokens are identical.
  AR greedily selects token `3831`, while MTP greedily selects token `95` from
  the same committed prefix.

The issue reproduces on a 255-token authorization-code fixture, but it is not a
generic chunk-boundary failure. Separate AR-derived continuation checks passed
all `128+1`, `128+127`, `128+128`, and `128+128+1` next-token comparisons under
both AR and MTP (24/24 total). MTP divergence also reproduces after a single
128-token prefill dispatch.

## Command or Request

Use the `authorization_code_255` prompt from
`tests/fixtures/deepseek_v4_chunked_prefill_accuracy_prompts.json` and run AR,
then K=1 MTP, serially on the same eight devices.

Relevant serving configuration:

```text
--model /data/models/dsv4-flash-w8a8
--backend npu
--platform a2a3
--devices 0,2,4,6,8,10,12,14
--dp 8 --ep 8 --tp 1
--block-size 128
--max-model-len 512
--max-num-seqs 32
--max-num-batched-tokens 512
--long-prefill-token-threshold 2048
--enable-chunked-prefill
--no-enable-prefix-caching
```

Generation settings:

```json
{
  "max_new_tokens": 64,
  "temperature": 0.0,
  "top_p": 1.0,
  "top_k": 0,
  "ignore_eos": true
}
```

For the MTP run, additionally use:

```text
--speculative-config '{"method":"mtp","num_speculative_tokens":1}'
```

Compare the returned token IDs exactly, rather than comparing only answer-level
text.

## Environment

| Component | Version |
|---|---|
| pypto-serving | `40ebf9a30b6a17a9f93e95396dffed9554137157` (PR #132 branch) |
| pypto-lib | `aea5bcba2489d6dc183ed35a5a784d8806e7ba4d` (PR #883 branch) |
| pypto | `71020585278b68f56c72c40d5570f07dbb20bc8b` |
| runtime | `3165cc89b6ea6b58a0bc01cbec2d5f72f2029c35` |
| pto-isa | `83d01313d9bfc247c4b7c8bcf969d1019f0d106f` |
| ptoas | `0.57` |
| CANN | `9.0.0` |
| Host | Linux aarch64 |
| Device / platform | eight A2/A3 devices: `0,2,4,6,8,10,12,14` |
| Model | DeepSeek V4 Flash W8A8, `/data/models/dsv4-flash-w8a8` |

## Logs

At the first divergence after correcting host committed-window publication:

```text
AR row0:
top_ids    = [3831, 95, 41653, 37236]
top_values = [32.758774, 32.471817, 29.725901, 27.758673]

MTP target row0:
top_ids    = [95, 3831, 41653, 37236]
top_values = [32.133121, 31.175119, 28.329367, 26.542364]
```

An earlier identical-prefix step also shows a much larger AR/MTP change for a
non-winning candidate:

```text
AR:
top_ids    = [7574, 7155, 9575, 12846]
top_values = [32.450748, 26.424942, 24.038696, 23.189779]

MTP target row0:
top_ids    = [7574, 7155, 7408, 12846]
top_values = [32.571129, 32.510063, 26.669779, 26.376173]
```

Relevant hardware tasks:

- `task_20260815_044128_115197417619`: async and sync MTP comparison;
  both initially diverged at index 21.
- `task_20260815_045502_130087127809`: AR/MTP top-4 logit capture; after the
  canonical host-window fix, visible divergence moved to index 48.
- `task_20260815_050449_141305231511`: temporary 8-row MTP tile; divergence
  still reproduced.

## Additional Context

DeepSeek MTP serving was introduced in PR #89. PR #120 later added arbitrary
draft depth and expanded/repacked MTP decode layouts. Both accuracy guards use
the short prompt `Huawei is` and require only this 10-token completion:

```text
 a leading global information and communications technology (ICT)
```

That guard passes, but it does not compare AR and MTP token streams and therefore
does not cover this losslessness requirement. PR #120 explicitly scopes the
device acceptance path to greedy-only sampling, so stochastic speculative
sampling is not expected here.

Suggested next checks:

1. Change only the S=2 row1 draft while holding row0 and its committed cache
   history fixed; row0 logits must remain unchanged.
2. Compare S=1 AR and S=2 MTP row0 hidden states layer by layer to find the first
   stage that diverges materially.
3. Inspect compressed-cache write/read ordering, row-wise causal metadata, raw
   KV page aliasing, and multi-active-token MoE dispatch/combine.
4. Add a deterministic CI regression that runs the same prompt under AR and K=1
   MTP and asserts exact token-ID equality for a sufficiently long decode.

Related work: PR #132 (chunked prefill) and pypto-lib PR #883.

**Comment by vegetabledoww** [comment]
阶段性进展记录（2026-08-15）：该问题暂时搁置，当前没有继续提交真机任务。

如果以“定位根因、完成修复、通过严格 AR/MTP 回归”为 100%，目前整体进度估计约 55%。已经完成稳定复现和较大范围排除，但尚未定位到具体出错的 layer/operator，也没有最终修复。

## 已经证实

1. 这不是每次运行随机变化。相同代码、配置、prompt 和 `temperature=0` 下，同步 MTP 重复运行可得到相同 token 序列。
2. MTP 与 AR 的逐 token 输出确实不等价。授权码用例可以稳定复现。
3. 问题不依赖跨 chunk prefill：
   - AR-derived continuation 的 `128+1`、`128+127`、`128+128`、`128+128+1` 下一 token 检查在 AR/MTP 下共 24/24 通过；
   - 只有一次 128-token prefill dispatch 的用例仍能出现 MTP 分叉。
4. 问题不是异步调度导致。任务 `task_20260815_044128_115197417619` 中，async/sync MTP 均复现分叉。
5. 问题不是 PR #120 将 MTP tile 从 T8 扩成 T16 导致。任务 `task_20260815_050449_141305231511` 临时恢复 T8 后仍然复现。
6. 找到一个独立的 host reclaim 问题：host 直接发布 `sampled_ids[:accepted_count]`，没有使用 device state 实际提交的 canonical committed window。本地改为发布 committed-window 后缀后，表面首次分叉从 0-based index 21 推迟到 index 48，说明该修改有效，但它没有解决底层 target logits 不等价问题。

上述 host reclaim 修改目前只是本地诊断性修改，尚未 commit/push，也不应描述成最终根因修复。

## 剩余的核心问题

在 canonical committed-window 修改后，AR/MTP 前 48 个生成 token 相同，但下一 token 仍然分叉：

```text
AR S=1 row0:
top_ids    = [3831, 95, 41653, 37236]
top_values = [32.758774, 32.471817, 29.725901, 27.758673]

MTP S=2 target row0:
top_ids    = [95, 3831, 41653, 37236]
top_values = [32.133121, 31.175119, 28.329367, 26.542364]
```

也观察到过同一已提交前缀下，MTP row0 中某个非 winning candidate 被明显抬高：

```text
AR:
top_ids    = [7574, 7155, 9575, 12846]
top_values = [32.450748, 26.424942, 24.038696, 23.189779]

MTP target row0:
top_ids    = [7574, 7155, 7408, 12846]
top_values = [32.571129, 32.510063, 26.669779, 26.376173]
```

这说明剩余问题已经进入 S=2 target model 的数值/状态路径，但“row0 被 row1 draft 污染”目前仍只是高优先级假设，尚未通过单变量实验直接证明。不要通过 epsilon tie-break 掩盖，因为差异不仅发生在接近平票的两个 token 上，候选 logit 的变化幅度也很大。

## 已检查但尚未定论的方向

- row-wise position、SWA/HCA/CSA metadata 表面上按每个 query row 独立构造；
- 仍需检查 compressor/indexer 是否先写入未来 row 的 state，再被 row0 读取；
- 仍需检查 raw KV 环形页/slot 是否存在 S=2 行别名；
- 仍需检查 MoE 在两个 active token 路由到同一 expert 时的 compact row、route 和 combine 回写；
- 需要做 layer-level hidden-state 二分，定位 S=1 AR 与 S=2 MTP row0 首次明显分叉的位置。

## PR 历史和测试盲区

- DeepSeek V4 MTP 最初由 Serving PR #89 引入；
- PR #120 是任意 draft depth 和 MTP layout/行打包扩展，不是最初引入；
- #89/#120 的 accuracy guard 都以短 prompt `Huawei is` 验证固定 10-token 文本，没有比较同一 prompt 的 AR/MTP token 序列，因此不能覆盖 greedy speculative decoding 的无损等价性要求。

## 后续恢复时建议的顺序

1. 先做 row1 draft 单变量实验：保持 row0 token、position 和已提交 cache 完全不变，只替换 row1 draft，比较 row0 logits 是否变化；
2. 若证实跨行影响，再比较 AR S=1 与 MTP S=2 的 row0 hidden state，逐 layer/stage 二分；
3. 修复后增加同 prompt、`temperature=0`、足够长输出的 AR/MTP exact-token CI guard；
4. 将本问题作为独立 MTP correctness PR 处理，不与 Chunk Prefill PR #132 继续混合。若根因在 pypto-lib，则配套提交 Lib PR，Serving PR 更新 submodule 并接入回归。

本阶段相关环境：Serving `40ebf9a30b6a17a9f93e95396dffed9554137157`，Lib `aea5bcba2489d6dc183ed35a5a784d8806e7ba4d`，设备 `0,2,4,6,8,10,12,14`。

**Comment by vegetabledoww** [comment]
补充并更正前一条评论中关于“重复性”的结论：此前“同步 MTP 重复运行可得到相同 token 序列”的表述过强，只能说明固定的一次同步单请求曾复现相同结果，不能据此声明 MTP K=1 在重复请求下是确定的。

当前更准确的实验结论如下：

1. **AR 顺序基线可重复。** 对小于 128 token 的同一 prompt，在 `temperature=0`、`top_p=1`、`top_k=0` 下，同步顺序请求 16/16、异步顺序请求 16/16 的 token IDs 完全一致。同步并发场景另有异常，本问题暂不把并发纳入讨论。
2. **MTP K=1 的重复性尚未通过。** 对 255-token 授权码 prompt，在没有并发、同一服务内顺序发送请求时，两轮结果中仍有部分请求分别从生成 token index 0、21、24、51 等位置开始分叉。因此目前不能说 MTP K=1 每次输出都相同。
3. **MTP 在确定性 greedy 条件下会偏离 AR。** 固定同步单请求的授权码用例中，AR 与 MTP 通常从第 22 个生成 token（0-based index 21）开始分叉；修正 host committed-window 后，分叉可以推迟到 index 48，但底层 target logits 仍然不等价。
4. **该问题不应归因于 Chunk Prefill。** 单个 128-token prefill chunk 即可复现 AR/MTP 分叉，AR-derived chunk continuation 的边界首 token 检查则为 24/24 通过。

基于以上证据，当前应将该问题定性为 **DeepSeek V4 MTP 路径的 correctness 回归/缺陷**。从功能责任边界看，应重点回溯最初引入 MTP 的 Serving PR #89，以及随后扩展 draft depth、layout 和行打包的 PR #120。两者现有测试都只检查短 prompt `Huawei is` 的固定 10-token 输出，没有验证 greedy 条件下 AR/MTP 的 exact-token 无损等价性，也没有覆盖同一服务内的重复请求。

目前尚未完成版本二分和 layer/operator 定位，所以还不能严格断言“第一个错误提交一定在 #89”或“由 #120 单独引入”。但可以明确：这不是 Chunk Prefill PR #132 需要承担的功能问题，而是 #89/#120 所建立和演进的 MTP 实现路径需要补齐正确性修复与回归测试。后续验收至少应同时满足：固定 replica 下 AR 自重复一致、MTP K=1 自重复一致、MTP K=1 与 AR 逐 token 一致。

**Comment by hashiqiqixian** [comment]
阶段性结论（2026-08-24）：基于当前官方 main 的小规模真机复测，暂时判断没有证据表明这里仍存在独立的 K=1 MTP 语义正确性问题。

本次使用以下版本，在 A2/A3 设备 8–15 上串行运行 AR 与 MTP K=1：

- pypto-serving: `013cabe853203f906d931089e3113b103fd439a3`
- pypto-lib: `2cd5f828bd3a54e2d073a7950b5fcfc0806b8bae`
- PyPTO: `97e4ef8ec422357c40ef54b1499511ecfb5ac3fb`
- simpler: `93adc3865e65fd72655f571f2e4165cf5ac02616`

小规模结果如下：

- `Huawei is`：AR 与 K=1 MTP 的 24 个生成 token 完全一致。
- 简单算术 `2+3`：token 序列不同，但两条路径均给出正确答案 `5`。
- 授权码提取：token 序列不同，但两条路径均正确返回 `CEDAR-4821`。
- 重复授权码请求：MTP K=1 自身的 token 序列不完全一致；同时 AR 自身的两次相同请求也不完全一致。

当前 pypto-lib 已通过 #974 恢复 `qkv_proj_rope` 的 split-K FP32 `AtomicAdd` 汇总。该实现按任务完成顺序累加，结果不保证 bit-reproducible；本次也观察到 AR 自身重复请求发生 token 波动。因此，目前的 AR/MTP exact-token 差异不能单独作为 MTP 路径错误的证据。

按当前“核心答案逻辑正确”的验收口径，本次 K=1 MTP smoke 暂时视为通过，阶段性判断没有独立 MTP 问题。这个结论不等价于完整精度验收，也不代表 greedy 逐 token 可重复性已经恢复；如果后续仍要求 speculative decoding 与确定性 AR 严格逐 token 等价，需要先恢复或提供一个确定性的 target-model reduction 基线，再重新执行严格 A/B。

本次未验证 K>1，也未进行完整精度矩阵。真机任务 `task_20260823_211658_241330213585` 正常结束（exit 0），设备锁均已释放。

**Comment by vegetabledoww** [comment]
实际上，吐字不同不是精度问题，只要符合逻辑，就是好的
